trieste/acquisition/rule.py

```diff
@@ -1097,6 +1104,225 @@ def state_func(
+    ):
+        """
+        Note that the optional parameters are set by a heuristic if not given by the user.
+
+        :param search_space: The search space.
+        :param num_query_points: The number of points in a batch.
+        :param L_min: Minimum allowed length of the trust region.
+        :param L_init: Initial length of the trust region.
+        :param L_max: Maximum allowed length of the trust region.
+        :param success_tolerance: Number of consecutive successes before changing region size.
+        :param failure tolerance: Number of consecutive failures before changing region size.
+        :param num_samples: Size of sample used for Thompson sampling.
+        """
+
+        if not num_query_points > 0:
+            raise ValueError(f"Num query points must be greater than 0, got {num_query_points}")
+
+        if num_query_points > 1:
+            raise NotImplementedError(
+                f"TURBO does not yet support batches greater than 1, got {num_query_points}"
+            )
+
+        # implement heuristic defaults for trust region params if not specified by user
+        if failure_tolerance is None:
+            failure_tolerance = math.ceil(search_space.dimension / num_query_points)
+        search_space_max_width = tf.reduce_max(search_space.upper - search_space.lower)
+        if L_min is None:
+            L_min = (0.5**7) * search_space_max_width
+        if L_init is None:
+            L_init = 0.8 * search_space_max_width
+        if L_max is None:
+            L_max = 1.6 * search_space_max_width
+        if num_samples is None:
+            num_samples = tf.clip_by_value(100 * search_space.dimension, 5_000, 1_000_000)
+
+        if not success_tolerance > 0:
+            raise ValueError(
+                f"success tolerance must be an integer greater than 0, got {success_tolerance}"
+            )
+        if not failure_tolerance > 0:
+            raise ValueError(
+                f"success tolerance must be an integer greater than 0, got {failure_tolerance}"
+            )
+        if not num_samples > 0:
+            raise ValueError(
+                f"""
+                number of samples for Thompson sampling must be an integer greater than
+                0, got {num_samples}
+                """
+            )
+
+        if L_min <= 0:
+            raise ValueError(f"L_min must be postive, got {L_min}")
+        if L_init <= 0:
+            raise ValueError(f"L_min must be postive, got {L_init}")
+        if L_max <= 0:
+            raise ValueError(f"L_min must be postive, got {L_max}")
+
+        self._num_query_points = num_query_points
+        self._L_min = L_min
+        self._L_init = L_init
+        self._L_max = L_max
+        self._success_tolerance = success_tolerance
+        self._failure_tolerance = failure_tolerance
+        self._num_samples = num_samples
+        self._thompson_sampler = ThompsonSamplerFromTrajectory(sample_min_value=False)
```

Comment:
there is no particular reason to get tied to TS, right? we could make a slightly more general TurBO that also works with EI, similar to TREGO?

Reply:
with the above base rule as default so it reverts to the original TurBO

Reply:
BTW this is not exactly in line with the original TurBO, it should be the `ExactThompsonSampler`, right?

Reply:
Ive allowed this to be custom